### PR TITLE
Add 0.5.0 of IDL

### PR DESCRIPTION
**Pull request by mtdowling**
Addresses #215

1. Only a single version statement can appear per/file. This was stated in the spec, but not enforced by the loader. Allowing only a single version statement per/file makes models simpler to parse and removes confusion if multiple statements appear.
2. Metadata must be defined after a version (if any) and before a namespace statement (if any). This change makes metadata appear in models more consistently and makes it easier to implement parsers.
3. Only a single namespace can appear per IDL file. Allowing only a single namespace makes models easier to parse and more consistent.
4. Use statements must appear after a namespace statement. This was the only way examples were presented in the spec, but it wasn't explicitly called out or enforced by the parser. This makes it much easier to understand the semantics of the use statement and makes it easier to implement parsers.

These changes are implemented in a backward compatible way. If a version of 0.4.0 is specified in a model explicitly, then none of the above constraints are enforced. If no version is set and any of the above constraints are violated, then a warning is emitted noting that the violation will not be allowed in future versions of Smithy. If a version of 0.5.0 is explicitly set, then the above violations will cause an ERROR and fail a model build.

### CODE REVIEW

File: docs/source/spec/core.rst [hunk 1]
```diff
@@ -2378,30 +2361,23 @@ Consider the following metadata definition:
...
 
-    namespace smithy.example
-
     metadata foo = {
-        MyString: MyString,
+        String: String,
     }
 
-    string MyString
+    string String
```
Comment (kstich) [inline docs/source/spec/core.rst]: Bit of a confusing example - is the metadata referencing `smithy.api#String` like described or this `string String`? If the former, can the `string String` be removed? If `string String` should stay, this needs a namespace.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelLoader.java [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+        } else if (next.lexeme.contains("'") || next.lexeme.contains("\"")) {
+            throw syntax("Unexpected syntax. Perhaps an unclosed quote?");
+        } else if (next.errorMessage == null) {
+            throw syntax("Unexpected syntax");
+        } else {
+            throw syntax(next.errorMessage);
+        }
+    }
+
+    private void parseService() {
```
Comment (kstich) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelLoader.java]: Nit: Reads a little oddly that these parse functions are between some of the token/lexer handling.

File: smithy-model/src/test/resources/software/amazon/smithy/model/loader/valid/version_0_4_0/multiple-versions-per-file.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+    "smithy": "0.5.0",
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/test/resources/software/amazon/smithy/model/loader/valid/version_0_4_0/multiple-versions-per-file.json]: No, these tests show how a .smithy file is loaded successfully into a .json file. We support loading multiple versions of the IDL and JSON AST format, but only support a single version for serializing models.

File: smithy-model/src/test/resources/software/amazon/smithy/model/loader/valid/version_0_4_0/use-before-namespace.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "smithy": "0.5.0",
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/test/resources/software/amazon/smithy/model/loader/valid/version_0_4_0/use-before-namespace.json]: No, these tests show how a .smithy file is loaded successfully into a .json file. We support loading multiple versions of the IDL and JSON AST format, but only support a single version for serializing models.

File: smithy-model/src/test/resources/software/amazon/smithy/model/loader/valid/version_0_4_0/metadata-can-come-before-namespace.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+    "smithy": "0.5.0",
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/test/resources/software/amazon/smithy/model/loader/valid/version_0_4_0/metadata-can-come-before-namespace.json]: See other comments

File: docs/source/spec/core.rst (hunk 1 above)
Comment (mtdowling) [reply docs/source/spec/core.rst]: Good catch. Removes the `string String`

**Comment by mtdowling** [comment]
Squashed and merged into add-0_5_0-format

**Pull request closed without merging**